pkg/io/v2/providers/providers.go

```diff
@@ -0,0 +1,215 @@
+	}
+
+	client, err := gcp.NewHTTPClient(
+		gcp.DefaultTransport(),
+		gcp.TokenSource(googleCredentials.TokenSource))
+	if err != nil {
+		return nil, fmt.Errorf("error creating GCP Http Client: %v", err)
+	}
+
+	bkt, err := gcsblob.OpenBucket(ctx, client, bucketName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error opening GCS bucket: %v", err)
+	}
+	return bkt, nil
+}
+
+// s3Credentials are credentials used to access S3 or an S3-compatible storage service
+// Endpoint is an optional property. Default is the AWS S3 endpoint. If set, the specified
+// endpoint will be used instead.
+type s3Credentials struct {
+	Region           string `json:"region"`
+	Endpoint         string `json:"endpoint"`
+	Insecure         bool   `json:"insecure"`
+	S3ForcePathStyle bool   `json:"s3_force_path_style"`
+	AccessKey        string `json:"access_key"`
+	SecretKey        string `json:"secret_key"`
+}
+
+// getS3Bucket opens a gocloud blob.Bucket based on given credentials in the format the
+// struct s3Credentials defines (see documentation of GetBucket for an example)
+func getS3Bucket(ctx context.Context, creds []byte, bucketName string) (*blob.Bucket, error) {
+	s3Credentials := &s3Credentials{}
+	if err := json.Unmarshal(creds, s3Credentials); err != nil {
+		return nil, fmt.Errorf("error getting S3 credentials from JSON: %v", err)
+	}
+
+	staticCredentials := credentials.NewStaticCredentials(s3Credentials.AccessKey, s3Credentials.SecretKey, "")
+
+	sess, err := session.NewSession(&aws.Config{
+		Credentials:      staticCredentials,
+		Endpoint:         aws.String(s3Credentials.Endpoint),
+		DisableSSL:       aws.Bool(s3Credentials.Insecure),
+		S3ForcePathStyle: aws.Bool(s3Credentials.S3ForcePathStyle),
+		Region:           aws.String(s3Credentials.Region),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error creating S3 Session: %v", err)
+	}
+
+	bkt, err := s3blob.OpenBucket(ctx, sess, bucketName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error opening S3 bucket: %v", err)
+	}
+	return bkt, nil
+}
+
+// ParseStoragePath parses storagePath and returns the storageProvider, bucket and relativePath
+// For example gs://prow-artifacts/test.log results in (gs, prow-artifacts, test.log)
+// Currently detected storageProviders are GS, S3 and file.
+// Paths with a leading / instead of a storageProvider prefix are treated as file paths for backwards
+// compatibility reasons.
+// File paths are split into a directory and a file. Directory is returned as bucket, file is returned
+// as relativePath.
+// For all other paths the first part is treated as storageProvider prefix, the second segment as bucket
+// and everything after the bucket as relativePath
+func ParseStoragePath(storagePath string) (storageProvider, bucket, relativePath string, err error) {
+	if strings.HasPrefix(storagePath, "gcs://") {
+		storagePath = strings.Replace(storagePath, "gcs://", "gs://", 1)
+	}
+	for _, spName := range storageProviders {
```

Comment:
There is no reason the whole known providers dance, it doesn't matter if we know the provider or not and we should not error if its unknown, because that puts us buck to square one of only supporting the provider we wrote code for. Instead we should:

* Replace the `gcs` prefix with `gcs` as we already do
* If `storagePath` contains no `://` string, prefix `storagePath` with `gs://`
* Parse the storagePath using a default scheme of `<<provider-name, delimited by ://>>://<< bucket name, delimited by string end or slash>>[/<< optional relative path >>]`
* Handle the special case of `file` if `storageProvider == file`
* Error if `bucket` is an empty string
* Return `storageProvider` (which may be one we don't know), `bucket` and `relativePath`

Reply:
Okay, how should we handle file paths. Currently file paths are specified like this `/tmp/path`. If I cannot handle it here. Every consumer has to detect if the path is a local path and has to add `file://`. Seem like that conflicts with: every path without prefix is a `gs://` path

I can only default to either file or gs. If i default to gs we have a breaking change for file paths

Reply:
Yeah it indeed does conflict. What consumers assume that unprefixed path is gs and what consumers assume its a file? If we have both, the consumer may need to do the defaulting.

Reply:
Right now (in this PR) we only have consumers assuming unprefixed paths are local files. I'm not sure right now if we have future consumers which assume the same for gcs. That's why I changed it

Reply:
Done